lec04-dplyr.Rmd

```diff
@@ -1,4 +1,724 @@
+```
+
+Now this looks good! There is a clear separation between different genus, but also significant spread within genus, for example in the weight of the green Neotoma observations. There are also two clearly separate clusters that are both colored in olive green (Dipodomys). Maybe separating the observations into different speciese would be better?
+
+```{r}
+ggplot(surveys_hf_wt, aes(x = weight, y = hindfoot_length, color = species)) +
+    geom_point(alpha = 0.2)
+```
+
+Great! Together with the genus plot, this definitely seem to explain most of the
+variance we see in the hindfoot length and weight measurements. It is still a
+bit messy as it appears like we have around 5 clusters, but there are 21 species
+in the legend.
+
+```{r}
+surveys %>% 
+    filter(!is.na(hindfoot_length) & !is.na(weight)) %>% 
+    group_by(species) %>% 
+    tally() %>% 
+    arrange(desc(n))
+```
+
+There is a big drop from 838 to 159, let's include only those with more than 800 observations.
+
+```{r}
+abundant_species <- surveys %>% 
+    filter(!is.na(hindfoot_length) & !is.na(weight)) %>% 
+    group_by(species) %>% 
+    tally() %>% 
+    arrange(desc(n)) %>% 
+    filter(n > 800) %>% # 
+    select(species)
+
+abundant_species
+```
+
+```{r}
+surveys_abun_species <- surveys %>% 
+    filter(!is.na(hindfoot_length) &
+        !is.na(weight) &
+        species %in% abundant_species$species)
+
+surveys_abun_species
+```
+
+Still has almost 25k observations, so only 10k was removed.
+
+```{r}
+ggplot(surveys_abun_species, aes(x = weight, y = hindfoot_length, color = species)) +
+    geom_point(alpha = 0.2)
+```
+
+
+> ### Challenge
+>
+> Create a scatter plot of `hindfoot_length` over `species` with the `weight` showing in different colors.
+> Is there any problem with this plot? *Hint: think about how many observations there are*
+
+<!--
+```{r}
+ggplot(surveys_abun_species, aes(x = weight, y = species, color = hindfoot_length)) +
+    geom_point(size = 0.1, position = 'jitter')
+```
+-->
+
+
+### Split-apply-combine... plot!
+
+In this section, we will learn how to work with `**dplyr**` and `**ggplot**`
+together. Aided by the pipes (`%>%`), we can create a powerul data exploration
+workflow using these two packages.
+
+Let's calculate number of counts per year for each species. First we need
+to group the data and count records within each group:
+
+```{r}
+surveys_abun_species %>%
+    group_by(year, genus) %>%
+    tally() %>% 
+    arrange(desc(n)) # Adding arrange just to compare with histogram
+```
+
+We could assign this table to a variable, and then pass that variable to `ggplot()`.
+
+```{r}
+yearly_counts <- surveys_abun_species %>%
+    group_by(year, species) %>%
+    tally()
+
+ggplot(yearly_counts, aes(x = n)) +
+    geom_histogram()
+```
+
+Creating an intermediate variable, would be prefereable if you are preferring a time consuming calculation, because you would not want to do that operation everytime you change some plot aesthetic.
```

Comment:
`Creating an intermediate variable would be preferable if it is a time consuming calculation`